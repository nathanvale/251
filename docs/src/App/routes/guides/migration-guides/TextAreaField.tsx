import React from "react";
import ReactMarkdown from "react-markdown";
import {
  Emphasis,
  List,
  ListItem,
  Stack,
  Strong,
  Text,
  TextLink,
} from "@origin-digital/ods-core";
import { DiffCodeBlock } from "../../../DiffCodeBlock/DiffCodeBlock";
import { Page, PageSection } from "../../../Page/Page";
import {
  baseMigrationGuidesURL,
  ComponentMigrationGuide,
  STYLE_GUIDE_URL,
} from "./consts";

const COMP_NAME = "TextAreaField";

export const textAreaFieldMigrationGuide: ComponentMigrationGuide = {
  compName: COMP_NAME,
  route: `${baseMigrationGuidesURL}/${COMP_NAME}`,
  ListItemComp: () => (
    <TextLink href={`${baseMigrationGuidesURL}/${COMP_NAME}`}>
      {COMP_NAME}
    </TextLink>
  ),
  PageComp: () => {
    const sections: PageSection[] = [
      {
        title: "Overview",
        children: (
          <Stack>
            <Text>
              As <Strong>TextAreaField</Strong> has most of its API in common
              with the <Strong>TextField</Strong> please refer to{" "}
              <TextLink href={`${baseMigrationGuidesURL}/TextField`}>
                TextField migration guide
              </TextLink>{" "}
              first, for shared props. Here we only provide migration guides
              that are unique to <Strong>TextAreaField</Strong>.
            </Text>
          </Stack>
        ),
      },
      {
        title: "API Changes",
        children: (
          <List>
            <ListItem>
              The <Emphasis>height</Emphasis> prop is removed. Use{" "}
              <Emphasis>(rows, rowsMax)</Emphasis> props instead. Contrary to{" "}
              <Emphasis>height</Emphasis>, <Emphasis>(rows, rowsMax)</Emphasis>{" "}
              are in the number of rows to show not pixels.
            </ListItem>
          </List>
        ),
      },
      {
        title: "Diff",
        children: (
          <ReactMarkdown
            renderers={{ code: DiffCodeBlock }}
            source={`\`\`\`diff
<TextAreaField
  ...
-  height={240}
+  rows={4}
+  rowmax={10}
/>

\`\`\``}
          />
        ),
      },
      {
        title: "Previous Implementation",
        children: (
          <List>
            <ListItem>
              <TextLink href={`${STYLE_GUIDE_URL}/${COMP_NAME}`}>
                Origin Style Guide
              </TextLink>
            </ListItem>
          </List>
        ),
      },
    ];

    return <Page title={`${COMP_NAME} Migration Guide`} sections={sections} />;
  },
};

export default textAreaFieldMigrationGuide;
