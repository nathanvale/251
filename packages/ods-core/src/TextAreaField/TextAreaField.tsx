import React from "react";
import { BaseFieldProps, TextFieldBase } from "../TextFieldBase/TextFieldBase";

export interface TextAreaFieldProps extends BaseFieldProps {
  domProps?: React.HTMLAttributes<HTMLTextAreaElement>;
  rows?: number | string;
  rowsMax?: number | string;
}

export const TextAreaField = ({ ...rest }: TextAreaFieldProps) => {
  return <TextFieldBase {...rest} multiline={true} />;
};

TextAreaField.displayName = "TextAreaField";
TextAreaField.defaultProps = {
  rows: 4,
  size: "medium",
  variant: "filled",
};
