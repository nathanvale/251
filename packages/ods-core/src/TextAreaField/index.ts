export { TextAreaField } from "./TextAreaField";
export type TextAreaFieldProps = import("./TextAreaField").TextAreaFieldProps;
