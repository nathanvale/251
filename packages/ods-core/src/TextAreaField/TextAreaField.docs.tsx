import React from "react";
import { ComponentDocs } from "@origin-digital/ods-types";
import { Stack } from "../Stack/Stack";
import { Text } from "../Text/Text";
import { baseFieldPropDesc } from "../TextFieldBase/TextFieldBase.docs";
import { TextAreaField, TextAreaFieldProps } from "./TextAreaField";

export const docs: ComponentDocs<TextAreaFieldProps> = {
  category: "Interaction",
  componentName: "TextAreaField",
  description: [
    "This is the underlying component of a number higher level Field components,",
    "such as TextField, SelectField, TextAreaField, etc.",
  ].join(" "),
  propDescriptions: {
    ...baseFieldPropDesc("TextField"),
    rows: `Number of rows to display. Default is 1.`,
    rowsMax: `Maximum number of rows to display. If the rows exceed this value, a vertical scrollbar appears.`,
    domProps:
      "The standard HTMLTextArea props. This will be directly passed down to the textarea element.",
  },
  migrationGuide: false,
  examples: {
    default: {
      Code: () => <TextAreaField id="taf-1" label="A text field" />,
    },
    additional: [
      {
        label: "rows and rowsMax",
        description: [
          "rows define the initial number of rows that the TextAreaField should take.",
          "rowsMax defines the maximum height of the TextAreaField.",
        ].join(" "),
        Code: () => (
          <TextAreaField id="taf-2" label="3 rows" rows="3" rowsMax="6" />
        ),
      },
      {
        label: "Handle change",
        description:
          "To handle change listen on onChange and provide the value prop",
        playroom: false,
        Code: () => {
          const [value, setValue] = React.useState<string>();
          return (
            <Stack space="medium">
              <TextAreaField
                id="text-field3"
                label="Name"
                helperText="Enter your name please"
                value={value}
                onChange={(e) => setValue(e.target.value)}
              />
              {value && <Text tone="positive">Entered name: {value}</Text>}
            </Stack>
          );
        },
        codeString: `const [value, setValue] = React.useState<string>();
return (<Stack space="medium">
  <TextAreaField
    id="text-field4"
    label="Name"
    helperText="Enter you name please"
    value={value}
    onChange={e => setValue(e.target.value)}
  />
  {value && <Text tone="positive">Entered name: {value}</Text>}
</Stack>)`,
      },
      {
        label: "Outlined variant",
        description: ["TextField also has an Outlined variant."].join(" "),
        Code: () => (
          <TextAreaField
            id="text-field11"
            label="Type here"
            variant="outlined"
            helperText="helper text for outlined input"
            placeholder="Placeholder"
            rows="5"
          />
        ),
      },
    ],
  },
  snippets: [
    {
      label: "Basic",
      Code: () => (
        <TextAreaField
          id="taf"
          label="label"
          placeholder="placeholder"
          helperText="details"
        />
      ),
    },
    {
      label: "rows and rowsMax",
      Code: () => (
        <TextAreaField id="taf" label="3 rows" rows="3" rowsMax="6" />
      ),
    },
  ],
};
