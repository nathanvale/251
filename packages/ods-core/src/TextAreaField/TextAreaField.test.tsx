import React from "react";
import {
  fireEvent,
  queryByAttribute,
  render,
} from "@origin-digital/ods-testing-library";
import { TextAreaField } from "./TextAreaField";

test("renders a textarea element", () => {
  const { getByTestId } = render(
    <TextAreaField id="my-taf" label="TextArea" />
  );

  const input = getByTestId("my-taf-input") as HTMLTextAreaElement;

  expect(input.tagName.toLowerCase()).toEqual("textarea");
});

test("renders multiple rows", () => {
  const { container } = render(
    <TextAreaField id="my-taf" label="TextArea" rows="3" rowsMax="6" />
  );

  expect(container).toMatchSnapshot();
});

test(`should render a TextAreaField in error state`, () => {
  const { container } = render(
    <TextAreaField
      id="my-text-area-field"
      label="The title"
      helperText="This is the info"
      error
    />
  );

  const label = queryByAttribute(
    "data-id",
    container,
    "my-text-area-field-label"
  );
  const helperText = queryByAttribute(
    "data-id",
    container,
    "my-text-area-field-helper-text"
  );

  expect(label).toHaveClass("Mui-error");
  expect(helperText).toHaveClass("Mui-error");
});

test("when a text is entered onChanged is called with the value of input", () => {
  const onChange = jest.fn();
  const { getByTestId } = render(
    <TextAreaField id="my-taf" onChange={onChange} />
  );

  const input = getByTestId("my-taf-input") as HTMLTextAreaElement;

  expect(input.value).toEqual("");

  fireEvent.change(input, { target: { value: "John" } });

  expect(onChange).toHaveBeenCalled();
  expect(input?.value).toEqual("John");
});
