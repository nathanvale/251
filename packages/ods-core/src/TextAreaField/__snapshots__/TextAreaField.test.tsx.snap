// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders multiple rows 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root textfieldbase-root-3 MuiFormControl-fullWidth"
    data-id="my-taf"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-filled"
      data-id="my-taf-label"
      data-shrink="false"
      for="my-taf"
      id="my-taf-label"
    >
      TextArea
    </label>
    <div
      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiFilledInput-multiline"
    >
      <textarea
        aria-describedby="my-taf-helper-text"
        aria-invalid="false"
        autocomplete="false"
        class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
        data-id="my-taf-input"
        id="my-taf"
        rows="3"
        style="height: 0px; overflow: hidden;"
      />
      <textarea
        aria-hidden="true"
        class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
        readonly=""
        style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); width: 100%;"
        tabindex="-1"
      />
    </div>
    <div
      class="MuiFormHelperText-root MuiFormHelperText-contained form-helper-text-grey400-colour-4"
      data-id="my-taf-helper-text"
      id="my-taf-helper-text"
    >
      <div>
        Â 
      </div>
    </div>
  </div>
</div>
`;
